### PR TITLE
feat(logger): for both controller level and component level

### PR DESCRIPTION
**Pull request by zdtsw**
## Description

ref: https://issues.redhat.com/browse/RHOAIENG-504

remove `Updating manifests : %v \n", manifestPath`
set two levels of logger
- build time: can update in the csv or ass flag to main binary: 
   `--log-level`: default to info, other valid value: debug, warning, error, panic and dpanic as standard logLevel
   `--log-mode`: default to devel, other valid value: prod
  when having debug/devel enabled, Events message will also show in the pod log
-  runtime:  add into DSCI devFlags, should not be something production needed, valid value: development, production, prod, devel.   or do not set .`spec.devFlags.logmode` at all which is using INFO
```
-spec:
  devFlags:
    logmode: development
```
if do not change anything config, users should get INFO level in general with development mode. which is the same as what we have now in controller. 
- by default (as not change any config):
  log: INFO level
  stacktrace: ERROR
- update DSCI devFlags.logmode to devel||development: for more logs
   log: INFO
   stacktrace: WARNING
- update DSCI devFlags.logmode to prod||production: for less logs
   log: WARNING
   stacktrace: ERROR

P.S:
this PR wont suspend warning from kustomize, e.g:
`# Warning: 'vars' is deprecated. Please use 'replacements' instead. [EXPERIMENTAL] Run 'kustomize edit fix' to update your Kustomization automatically.`

## How Has This Been Tested?

live build quay.io/wenzhou/opendatahub-operator:2.9.837-518

controller:
old:
`2024-02-02T09:02:32Z	INFO	controllers.DataScienceCluster	Reconciling DataScienceCluster resources	{"Request.Name": "default-dsc"}`
new without extra config:
`{"level":"info","ts":"2024-03-18T16:44:53Z","logger":"opendatahub.controllers.DataScienceCluster","msg":"Reconciling DataScienceCluster resources","Request.Name":"default-dsc"}`

component:
old:
```
2024-02-02T12:22:12Z	ERROR	controllers.DataScienceCluster	failed to reconcile ray on DataScienceCluster	{"instance.Name": "default-dsc", "error": "metadata.managedFields must be nil"}
github.com/opendatahub-io/opendatahub-operator/v2/controllers/datasciencecluster.(*DataScienceClusterReconciler).reportError
	/workspace/controllers/datasciencecluster/datasciencecluster_controller.go:302
github.com/opendatahub-io/opendatahub-operator/v2/controllers/datasciencecluster.(*DataScienceClusterReconciler).reconcileSubComponent
	/workspace/controllers/datasciencecluster/datasciencecluster_controller.go:271
github.com/opendatahub-io/opendatahub-operator/v2/controllers/datasciencecluster.(*DataScienceClusterReconciler).Reconcile
	/workspace/controllers/datasciencecluster/datasciencecluster_controller.go:208
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
```
new: 
```
{"level":"error","ts":"2024-03-18T17:10:19Z","logger":"opendatahub.controllers.DataScienceCluster","msg":"failed to reconcile codeflare on DataScienceCluster","instance.Name":"default-dsc","error":"must build at directory: not a valid directory: evalsymlink failure on '/opt/manifests/codeflare/default/default' : lstat /opt/manifests/codeflare: no such file or directory","stacktrace":"github.com/opendatahub-io/opendatahub-operator/v2/controllers/datasciencecluster.(*DataScienceClusterReconciler).reportError\n\t/workspace/controllers/datasciencecluster/datasciencecluster_controller.go:314\ngithub.com/opendatahub-io/opendatahub-operator/v2/controllers/datasciencecluster.(*DataScienceClusterReconciler).reconcileSubComponent\n\t/workspace/controllers/datasciencecluster/datasciencecluster_controller.go:281\ngithub.com/opendatahub-io/opendatahub-operator/v2/controllers/datasciencecluster.(*DataScienceClusterReconciler).Reconcile\n\t/workspace/controllers/datasciencecluster/datasciencecluster_controller.go:215\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235"}
```
## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/codeflare/codeflare.go [hunk 1]
```diff
@@ -83,15 +86,17 @@ func (c *CodeFlare) ReconcileComponent(ctx context.Context, cli client.Client, r
 		}
 
 		if found, err := deploy.OperatorExists(cli, dependentOperator); err != nil {
-			return fmt.Errorf("operator exists throws error %w", err)
+			logger.Error("operator exists throws error", zap.Error(err))
```
Comment (aslakknutsen) [inline components/codeflare/codeflare.go:89]: Handle or return, not both. Wrap the error if it needs local context.

File: components/component.go [hunk 2]
```diff
@@ -87,6 +89,30 @@ type ComponentInterface interface {
...
 	OverrideManifests(platform string) error
 	UpdatePrometheusConfig(cli client.Client, enable bool, component string) error
+	ConfigLogger(dscispec *dsciv1.DSCInitializationSpec) *zap.Logger
+}
+
+// by default, ConfigLogger set InfoLevel and Caller to InfoLevel, Stacktrace to ErrorLevel (close to what we have now)
+// config to use devel/development: change to DebugLevel, with Caller to DebugLevel, keep stacktrace to ErrorLevel
+// config to use prod/production: change to ErrorLevel, with Caller to ErrorLevel and keep stracetrace to ErrorLevel
+// not set a "default" in switch because support value in LogMode is enum.
+func (c *Component) ConfigLogger(dscispec *dsciv1.DSCInitializationSpec) *zap.Logger {
```
Comment (aslakknutsen) [inline components/component.go:99]: Why the move to using the zap Logger directly? It was already used behind logr and logr is used everywhere else...?

File: components/dashboard/dashboard.go [hunk 3]
```diff
@@ -98,6 +99,13 @@ func (d *Dashboard) ReconcileComponent(ctx context.Context,
 		return err
 	}
 
+	var logger *zap.Logger
```
Comment (aslakknutsen) [inline components/dashboard/dashboard.go:102]: Why not pass down the Reonciler logr intance?

File: components/trustyai/trustyai.go [hunk 4]
```diff
@@ -57,6 +58,8 @@ func (t *TrustyAI) ReconcileComponent(ctx context.Context, cli client.Client, re
 		"trustyaiServiceImage":  "RELATED_IMAGE_ODH_TRUSTYAI_SERVICE_IMAGE",
 		"trustyaiOperatorImage": "RELATED_IMAGE_ODH_TRUSTYAI_SERVICE_OPERATOR_IMAGE",
 	}
+	logger := t.ConfigLogger(dscispec).With(zap.String("component", ComponentName))
```
Comment (aslakknutsen) [inline components/trustyai/trustyai.go:61]: logr.WithValues("component", ComponentName).Info...?

File: components/component.go [hunk 5]
```diff
@@ -80,13 +83,52 @@ type ManifestsConfig struct {
...
 	SetImageParamsMap(imageMap map[string]string) map[string]string
 	OverrideManifests(platform string) error
 	UpdatePrometheusConfig(cli client.Client, enable bool, component string) error
+	ConfigLogger(logger logr.Logger, component string, dscispec *dsciv1.DSCInitializationSpec) logr.Logger
+}
+
+// by default, ConfigLogger with jsonEncoder, logLevel=Info, stackTraceLevel=Error
+// when set in DSCI: devel Mode defaults(encoder=consoleEncoder,logLevel=Debug,stackTraceLevel=Info with lower case for key
+// when set in DSCI: prod  Mode defaults(encoder=jsonEncoder,   logLevel=Info, stackTraceLevel=Error with upper case for key.
+func (c *Component) ConfigLogger(logger logr.Logger, component string, dscispec *dsciv1.DSCInitializationSpec) logr.Logger {
```
Comment (aslakknutsen) [inline components/component.go:100]: Could we unify how the loggers are created between here and main.go?

Comment (zdtsw) [reply components/component.go:95]: did some updates, not sure if this is what you meant for "unify" ?

File: components/codeflare/codeflare.go (hunk 1 above)
Comment (zdtsw) [reply components/codeflare/codeflare.go]: ERROR are all set to only print and handle by DSC's logger,  
component level only INFO used by its logger

File: components/component.go (hunk 5 above)
Comment (aslakknutsen) [reply components/component.go:95]: unify = reuse code / call same function. Make a change in 1 location. Ensure they behave the same etc.. :)

Comment (zdtsw) [reply components/component.go:95]: in this case, it cannot be a method for DSC, gonna make it is util function

**Comment by VaishnaviHire** [comment]
Regarding the customize warnings, I think components will need to move to v5 of customize, however this will not be possible if upstream version of the components have not yet updated to v5. 

There is also a fix PR for this https://github.com/kubernetes-sigs/kustomize/pull/5466 that we can utilize once merged

**Review by VaishnaviHire — COMMENTED** [review]
/lgtm

**Review by VaishnaviHire — APPROVED** [review]
/lgtm

**Pull request merged**